purchase_work_acceptance/models/account_invoice.py

```diff
@@ -0,0 +1,169 @@
+# Copyright 2019 Ecosoft Co., Ltd. (http://ecosoft.co.th)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models, _
+from odoo.exceptions import ValidationError
+from odoo.tools.float_utils import float_compare
+
+
+class AccountInvoice(models.Model):
+    _inherit = 'account.invoice'
+
+    wa_id = fields.Many2one(
+        comodel_name='work.acceptance',
+        string='WA Reference',
+        domain=[('state', '=', 'accept')],
+        store=True,
+    )
+
+    @api.multi
+    def action_invoice_open(self):
+        self.ensure_one()
+        if self.wa_id:
+            wa_lines = self.env['work.acceptance.line']
+            for line in self.invoice_line_ids:
+                id = line.purchase_line_id.id
+                wa_line = wa_lines.search([('purchase_line_id', '=', id)])
+                if line.quantity > wa_line.product_qty:
+                    raise ValidationError(_('You cannot validate a bill if '
+                                          'Quantity more than accepted quantity'))
+        return super(AccountInvoice, self).action_invoice_open()
+
+    def _prepare_invoice_line_from_wa_line(self, line):
```

Comment:
From this point on, I think it need to be redo in different way. From what I see (correct if I am wrong), you are trying to use WA's line to create invoice line.

**WA line should not be used to create invoice line**

Invoice line should be created as as normal. But during it creation, just try to snap in and change the quantity and price_unit as in WA.

WA Line has purchase_line_id, and invoice_line also has purchase_line_id, so it should be easy to do so. In the method _prepare_invoice_line_from_po_line() which is now used to prepare invoice line, inherit it, it is easy to change the quantity and price_unit.

Note: it is super important that, the addon we create, do not **replace** the existing functionality, but only **extend* it.